src/module.c

```diff
@@ -807,11 +806,8 @@ void jl_binding_deprecation_warning(jl_module_t *m, jl_binding_t *b)
 JL_DLLEXPORT void jl_checked_assignment(jl_binding_t *b, jl_value_t *rhs)
 {
     jl_value_t *old_ty = NULL;
-    if (!jl_atomic_cmpswap_relaxed(&b->ty, &old_ty, (jl_value_t*)jl_any_type) && !jl_isa(rhs, old_ty)) {
-        jl_errorf("cannot assign an incompatible value to the global %s.",
-                  jl_symbol_name(b->name));
-    }
     if (b->constp) {
+        jl_atomic_cmpswap_relaxed(&b->ty, &old_ty, jl_typeof(rhs));
```

Comment:
Same above.

Reply:
I'm pretty sure right now we always assume that a non-null value implies that the binding type is also set, while setting a type for an uninitialized binding is perfectly fine. In case of  a race we might have set the wrong type, but this should then be caught and error appropriately before we set the value to something inconsistent. Or is there another reason for doing it the other way around?

Reply:
I forgot to add another check of the binding type here though, I can add that. (Although declaring constants isn't really thread-safe right now anyways.)

Reply:
I was just thinking of making setting the value and type more atomic, but I guess the behavior ends up being the same either way.